File: Chapter 10/Python/section_10_3_resilience_decorators/resilient/core/application/checkout_orchestrator.py
import uuid
from ..domain.order_status import OrderStatus
from ..ports.payment_gateway import PaymentGateway
from ..ports.message_queue import MessageQueue

class CheckoutOrchestrator:
    """
    THE CORE APPLICATION LAYER:
    
    DESIGN NOTE:
    This is the "Brain" of the hexagon. It coordinates the business flow 
    using Ports. It doesn't know about 'tenacity' or 'requests'; it only 
    knows the 'Policy' for success and the 'Plan B' for failure.
    """
    def __init__(self, payment_port: PaymentGateway, queue_port: MessageQueue):
        self.payment_port = payment_port
        self.queue_port = queue_port

    def process_checkout(self, order_id: str, amount: float) -> OrderStatus:
        # Idempotency Key is a Business Concern
        idempotency_key = str(uuid.uuid4())
        
        try:
            # 1. THE HAPPY PATH
            self.payment_port.charge(amount, order_id, idempotency_key)
            print("      [Infrastructure] Transaction successful.")
            return OrderStatus.PAID
            
        except Exception:
            # THE FALLBACK (Plan B)
            # When the Adapter's retries fail, the Orchestrator pivots.
            print(f"      [FALLBACK] PRIMARY FAILED. Executing Plan B (Queueing).")
            
            self.queue_port.enqueue({
                "order_id": order_id,
                "amount": amount,
                "status": OrderStatus.PENDING_PAYMENT.value,
                "idempotency_key": idempotency_key
            })
            return OrderStatus.PENDING_PAYMENT

File: Chapter 10/C#/Section_10_3_Resilience_Decorators/Resilient/Core/Application/CheckoutOrchestrator.cs
using System;
using System.Threading.Tasks;
using Chapter10.Resilient.Core.Domain;
using Chapter10.Resilient.Core.Ports;

namespace Chapter10.Resilient.Core.Application
{
    /// <summary>
    /// THE CORE APPLICATION LAYER:
    /// * @description
    /// This orchestrator coordinates the business flow using Ports. It 
    /// manages the "What" (Business Policy) while remaining blind 
    /// to the "How" (Infrastructure Implementation).
    ///
    /// ARCHITECTURAL CRITIQUE:
    /// 1. PIVOT ON FAILURE: When the primary gateway's internal retries are 
    ///    exhausted, the Orchestrator executes 'Plan B' (The Queue).
    /// 2. IDEMPOTENCY: The key is generated in the Core. This ensures that 
    ///    if a message is processed later from the queue, we don't 
    ///    double-charge the customer.
    /// </summary>
    public class CheckoutOrchestrator
    {
        private readonly IPaymentGateway _paymentPort;
        private readonly IMessageQueue _queuePort;

        public CheckoutOrchestrator(IPaymentGateway paymentPort, IMessageQueue queuePort)
        {
            _paymentPort = paymentPort;
            _queuePort = queuePort;
        }

        public async Task<OrderStatus> ProcessCheckout(string orderId, decimal amount)
        {
            // Idempotency generation is a Core Business concern.
            string idempotencyKey = Guid.NewGuid().ToString();

            try
            {
                // 1. THE HAPPY PATH (Hidden retries happen inside the adapter)
                await _paymentPort.ChargeAsync(amount, orderId, idempotencyKey);
                Console.WriteLine("      [Core Application] PRIMARY SUCCESS: Transaction PAID.");
                return OrderStatus.Paid;
            }
            catch (Exception ex)
            {
                // 2. THE FALLBACK (Plan B)
                Console.WriteLine($"      [Core Application] PRIMARY FAILED: {ex.Message}");
                Console.WriteLine("      [Core Application] EXECUTING PLAN B: Securing data in Queue.");

                // We bundle the data into a unified payload for the Port
                var payload = new
                {
                    OrderId = orderId,
                    Amount = amount,
                    Status = OrderStatus.PendingPayment.ToString(),
                    IdempotencyKey = idempotencyKey,
                    QueuedAt = DateTime.UtcNow
                };

                await _queuePort.Enqueue(payload);

                return OrderStatus.PendingPayment;
            }
        }
    }
}